xrootd/protocol/protocol/protocol.go

```diff
@@ -0,0 +1,121 @@
+
+// GeneralResponseLength is the length of GeneralResponse in bytes.
+const GeneralResponseLength = 8
+
+// General response is the response that is always returned from xrootd server.
+// It contains protocol version and flags that describe server type.
+type GeneralResponse struct {
+	BinaryProtocolVersion int32
+	Flags                 Flags
+}
+
+// Flags are the flags that define xrootd server type. See xrootd protocol specification for further info.
+type Flags int32
+
+const (
+	IsServer     Flags = 0x00000001
+	IsManager    Flags = 0x00000002
+	IsMeta       Flags = 0x00000100
+	IsProxy      Flags = 0x00000200
+	IsSupervisor Flags = 0x00000400
+)
+
+type SecurityOptions byte
+
+const (
+	// None specifies that no security options are provided.
+	None SecurityOptions = 0
+	// ForceSecurity specifies that signing is required even if the authentication
+	// protocol does not support generic encryption.
+	ForceSecurity SecurityOptions = 0x02
+)
+
+// SecurityInfoLength is the length of SecurityInfo in bytes.
+const SecurityInfoLength = 6
+
+// SecurityInfo is a response part that is provided when required (if server supports that).
+// It contains security version, security options, security level, and number of following security overrides, if any.
+type SecurityInfo struct {
+	Reserved1             byte
+	Reserved2             byte
+	SecurityVersion       byte
+	SecurityOptions       SecurityOptions
+	SecurityLevel         byte
+	SecurityOverridesSize byte
+}
+
+// RequestLevel is the security requirement that the associated request is to have.
+type RequestLevel byte
+
+const (
+	// SignNone indicates that the request need not to be signed.
+	SignNone RequestLevel = 0
+	// SignLikely indicates that the request must be signed if it modifies data.
+	SignLikely RequestLevel = 1
+	// SignNeeded indicates that the request mush be signed.
+	SignNeeded RequestLevel = 2
+)
+
+// SecurityOverrideLength is the length of SecurityOverride in bytes.
+const SecurityOverrideLength = 2
+
+// SecurityOverride is a alteration needed to the specified predefined security level.
+// It consist of the request index and the security requirement that associated request is to have.
+// Request index is calculated as:
+//     (request code) - (request code of Auth request)
+// according to xrootd protocol specification.
+type SecurityOverride struct {
+	RequestIndex byte
+	RequestLevel RequestLevel
+}
+
+// RequestOptions specifies what should be returned as part of response.
+type RequestOptions byte
+
+const (
+	// RequestOptionsNone specifies that only general response should be returned.
+	RequestOptionsNone RequestOptions = 0
+	// ReturnSecurityRequirements specifies that security requirements should be returned
+	// if that's supported by the server.
+	ReturnSecurityRequirements RequestOptions = 1
+)
+
+// Request is the struct that holds protocol request parameters.
+type Request struct {
+	ClientProtocolVersion int32
+	Reserved1             [11]byte
```

Comment:
if it works with the current `reflect`-based code of the marshaling/unmarshaling protocol, we should rename these `ReservedX` fields into just `_`.

if this doesn't work with the `reflect`-based code, then leave it like that but add a FIXME stating we should rename these into `_` when the automatically generated (un)marshaling code is in place.